pkg/util/taskgroup/taskgroup.go

```diff
@@ -0,0 +1,109 @@
+// Originally taken from https://github.com/ptxmac/multierrgroup
+
+// Package taskgroup provides a mix of multierr and errgroup
+// See documentation for https://pkg.go.dev/go.uber.org/multierr and https://pkg.go.dev/golang.org/x/sync/errgroup
+package taskgroup
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+// Group manages goroutines and collect all the errors.
+// See https://pkg.go.dev/golang.org/x/sync/errgroup#group for more information
+type Group interface {
+	WithContext(ctx context.Context) context.Context
+	WithPanicHandler(f func(any))
+	Wait() error
+	Go(name string, f func(logger *zap.Logger) error)
+}
+
+type group struct {
+	cancel       context.CancelFunc
+	logger       *zap.Logger
+	panicHandler func(any)
+
+	wg sync.WaitGroup
+
+	errMutex sync.Mutex
+	err      error
+}
+
+// NewGroup returns a new Group.
+func NewGroup(logger *zap.Logger) Group {
+	return &group{
+		cancel:       nil, // Set separately by WithContext
+		panicHandler: nil, // Set separately by WithPanicHandler
+		logger:       logger,
+		wg:           sync.WaitGroup{},
+
+		errMutex: sync.Mutex{},
+		err:      nil,
+	}
+}
+
+// WithContext updates the current Group with a associated Context.
+// The context will be canceled if any goroutine returns an error.
+// See https://pkg.go.dev/golang.org/x/sync/errgroup#WithContext
+func (g *group) WithContext(ctx context.Context) context.Context {
+	ctx, g.cancel = context.WithCancel(ctx)
+	return ctx
+}
+
+// WithPanicHandler sets a panic handler for the group.
+func (g *group) WithPanicHandler(f func(any)) {
+	g.panicHandler = f
+}
```

Comment:
In the past, I've seen `WithFoo` methods typically return a new version of the callee, with "foo" added. Maybe these should be named more like `SetFoo` instead?

That said: Is there a typical use case for adding a context after construction? IMO the API would be cleaner if it's an optional arg to `NewGroup`, plus a new function to fetch the context for the `Group`

Reply:
> IMO the API would be cleaner if it's an optional arg
Right, implemented in 836b1b0

This has and added advantage: no need to specify context if Background is the parent
And a disadvantage: the API is different from the one which is usually used around contexts: `ctx, ... := func1(ctx)`